core/src/main/scala/chisel3/reflect/DataMirror.scala

```diff
@@ -194,11 +194,66 @@ object DataMirror {
     }
   }
 
-  /** Return all expanded components, including intermediate aggregate nodes
+  // Old definition of collectLeafMembers
+  @deprecated("Use DataMirror.collectLeafMembers instead")
+  def getLeafs(d: Data): Seq[Data] = collectLeafMembers(d)
+
+  // Old definition of collectAllChildren
+  @deprecated("Use DataMirror.collectAllMembers instead")
+  def getIntermediateAndLeafs(d: Data): Seq[Data] = collectAllMembers(d)
+
+  /** Recursively collect just the leaf components of a data component's children
+    * (i.e. anything that isn't a `Record` or a `Vec`, but an `Element`)
+    *
+    * @param d Data component to recursively collect leaf components.
+    *
+    * @return All `Element` components; intermediate fields/indices are not included
+    */
+  def collectLeafMembers(d: Data): Seq[Data] =
+    DataMirror.collectMembers(d) { case x: Element => x.asInstanceOf[Data] }.toSeq
+
+  /** Recursively collect all expanded member components of a data component, including
+    * intermediate aggregate nodes
     *
-    * @param d Component to find leafs if aggregate typed. Intermediate fields/indicies ARE included
+    * @param d Data component to recursively collect components.
+    *
+    * @return All member components; intermediate fields/indices ARE included
     */
-  def getIntermediateAndLeafs(d: Data): Seq[Data] = collectMembers(d) { case x => x }.toSeq
+  def collectAllMembers(d: Data): Seq[Data] = collectMembers(d) { case x => x }.toSeq
+
+  /** Recursively collects all fields selected by collector within a data and additionally generates
+    * path names for each field
+    * Accepts a collector partial function, rather than a collector function
+    *
+    * @param data Data to collect fields, as well as all children datas it directly and indirectly instantiates
+    * @param path Recursively generated path name, starting with a root path
+    * @param collector Collector partial function to pick which components to collect
+    *
+    * @return A sequence of pairs that map a data field to its corresponding path name
+    *
+    * @tparam T Type of the component that will be collected
+    */
+  private[chisel3] def collectMembersAndPaths[T](
+    d:         Data,
+    path:      String = ""
+  )(collector: PartialFunction[Data, T]
+  ): Iterable[(T, String)] = {
+    val myItems = collector.lift(d).map { x => (x -> path) }
+    val deepChildrenItems = d match {
+      case a: Record =>
+        a._elements.flatMap {
+          case (fieldName, fieldData) =>
+            collectMembersAndPaths(fieldData, s"$path.$fieldName")(collector)
+        }
+      case a: Vec[_] =>
+        a.elementsIterator.zipWithIndex.flatMap {
+          case (fieldData, fieldIndex) =>
+            collectMembersAndPaths(fieldData, s"$path($fieldIndex)")(collector)
+        }
+      case other => Nil
+    }
+    myItems ++ deepChildrenItems
+  }
```

Comment:
```suggestion
  ): Iterable[(T, String)] = new Iterable {
    def iterator = {
      val myItems = collector.lift(d).map { x => (x -> path) }
      val deepChildrenItems = d match {
        case a: Record =>
          a._elements.iterator.flatMap {
            case (fieldName, fieldData) =>
              collectMembersAndPaths(fieldData, s"$path.$fieldName")(collector)
          }
        case a: Vec[_] =>
          a.elementsIterator.zipWithIndex.flatMap {
            case (fieldData, fieldIndex) =>
              collectMembersAndPaths(fieldData, s"$path($fieldIndex)")(collector)
          }
        case other => Nil
      }
      myItems.iterator ++ deepChildrenItems
    }
  }
```
This suggestion is mainly tabbing things over but it also adds `.iterator` after `a_.elements` in the `Record` block.

It's nice that `Option.++` returns `Iterable`, but concretely it just creates a `List` which is counterproductive for the use case of doing this lazily.

This suggestion is a bit subtle but the idea is that we can do this with iterators which are lazily and extremely efficient and just wrap it in a `new Iterable` so that the API is safely immutable (each iteration of the returned Iterable will create new single-use iterators).

I also suggest making this change below to `collectMembers`.